msal/application.py

```diff
@@ -1376,11 +1379,11 @@ def _acquire_token_silent_from_cache_and_possibly_refresh_it(
                 query["key_id"] = key_id
             now = time.time()
             refresh_reason = msal.telemetry.AT_ABSENT
-            for entry in self.token_cache._find(  # It returns a generator
+            for entry in self.token_cache.search(  # We want a generator here for performance, so that we can break early in cache-hit
                 self.token_cache.CredentialType.ACCESS_TOKEN,
                 target=scopes,
                 query=query,
-            ):  # Note that _find() holds a lock during this for loop;
+            ):  # Note that search() holds a lock during this for loop;
```

Comment:
`TokenCache` uses a `RLock`:

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/c4152d2920f16c0416ebcfa1b88c34d862db56e3/msal/token_cache.py#L42

This only prevents the enclosed code from being executed by multiple threads:

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/c4152d2920f16c0416ebcfa1b88c34d862db56e3/msal/token_cache.py#L142-L151

However, it doesn't prevent deleting an element from the token cache as `RLock` allows multiple acquisition by the same thread.

Reply:
[Discussed here](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/693#discussion_r1621227729)